streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java

```diff
@@ -62,24 +75,60 @@ public void addChild(ProcessorNode<?, ?> child) {
         children.add(child);
     }
 
+
     public void init(ProcessorContext context) {
+        long startNs = time.nanoseconds();
         try {
             processor.init(context);
         } catch (Exception e) {
             throw new StreamsException(String.format("failed to initialize processor %s", name), e);
         }
+        this.nodeMetrics = new NodeMetrics(context.metrics(), name,  "task." + context.taskId());
+        this.nodeMetrics.metrics.recordLatency(nodeMetrics.nodeCreationSensor, startNs, time.nanoseconds());
     }
 
     public void close() {
+        long startNs = time.nanoseconds();
         try {
             processor.close();
+            nodeMetrics.metrics.recordLatency(nodeMetrics.nodeDestructionSensor, startNs, time.nanoseconds());
+            nodeMetrics.removeAllSensors();
         } catch (Exception e) {
             throw new StreamsException(String.format("failed to close processor %s", name), e);
         }
     }
 
     public void process(final K key, final V value) {
+        long startNs = -1;
+        // expensive operation. Only perform if we are recording this metric
+        if (nodeMetrics.nodeProcessTimeSensor.maybeRecord()) {
```

Comment:
In this method and `punctuate` the blocks of code are largely the same. Is this pattern going to be common? if it is just in this class, then i'd probably create a method in here that accepts a `Runnable` and have `process` and `punctuate` delegate to it, i.e., 

```
private void measureLatency(final Runnable action, final Sensor sensor) {
   long startNs = -1;
   if (nodeMetrics.nodeProcessTimeSensor.maybeRecord()) {
       startNs = time.nanoseconds();
   }
  action.run();
  if (startNs != -1) {
      nodeMetrics.metrics.recordLatency(sensor, startNs, time.nanoseconds());
  }
}
```